kv: generate debug output on marshaling panics

tbg opened the pull request:
We're seeing panics from within gogoproto marshaling that indicate that
we're mutating a protobuf while it's being marshaled (thus increasing
the size necessary to marshal it, which is not reflected in the slice
being marshaled into).

Since we haven't managed to figure this out just by thinking hard, it's
time to add some debugging into the mix.

Since this hasn't popped up during our testrace builds, I assume it's
either rare enough or just not tickled in any of the tests (many of
which don't even run under `race` because things get too slow).

My hope is that looking at the bytes we get ouf of this logging we'll
see something that looks out of place and can trace it down.

Touches #34241.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35202)
<!-- Reviewable:end -->

jordanlewis reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> good idea

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35202#-:-LZdwaM29kktAWQBg9y7:bnaseej)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

tbg commented:
TFTR!

bors r=jordanlewis

tbg committed:
kv: generate debug output on marshaling panics

We're seeing panics from within gogoproto marshaling that indicate that
we're mutating a protobuf while it's being marshaled (thus increasing
the size necessary to marshal it, which is not reflected in the slice
being marshaled into).

Since we haven't managed to figure this out just by thinking hard, it's
time to add some debugging into the mix.

Since this hasn't popped up during our testrace builds, I assume it's
either rare enough or just not tickled in any of the tests (many of
which don't even run under `race` because things get too slow).

My hope is that looking at the bytes we get ouf of this logging we'll
see something that looks out of place and can trace it down.

Touches #34241.

Release note: None

tbg commented:
Oops, had triggered too early. Linter is happy now.

bors r=jordanlewis
